src/resmom/mom_comm.c

```diff
@@ -6156,14 +6525,459 @@ tm_request(int fd, int version)
+				free(tempnid);
+				free(streamlist);
+				goto err;
+			}
+
+			/* go through the list of nodes and send a request */
+			for (i = 0; i < list_size; i++) {
+				/* reset for each node in the list */
+				alreadyadded = 0;
+				tvnodeid = node_list[i];
+				/* check the node number is legal */
+				pnode = pjob->ji_vnods;
+				for (k = 0; k < pjob->ji_numvnod; k++, pnode++) {
+					if (pnode->vn_node == tvnodeid)
+						break;
+				}
+				if (k == pjob->ji_numvnod) {
+					sprintf(log_buffer, "node %d not found", tvnodeid);
+					log_joberr(-1, __func__, log_buffer, jobid);
+                			ret = tm_reply(fd, version, TM_ERROR, event);
+			                if (ret != DIS_SUCCESS) {
+						free(node_list);
+						free(tempinfo);
+						free(temptaskid);
+						free(tempnid);
+						free(streamlist);
+                       				goto done;
+					}
+			                ret = diswsi(fd, TM_ENOTFOUND);
+			                if (ret != DIS_SUCCESS) {
+						free(node_list);
+						free(tempinfo);
+						free(temptaskid);
+						free(tempnid);
+						free(streamlist);
+                       				goto done;
+					}
+        			}
+				phost = pnode->vn_host;
+				
+				/*
+				 ** If the spawn happens on me, just do it.
+				 */
+				if (pjob->ji_nodeid == TO_PHYNODE(tvnodeid)) {
+#ifdef PMIX
+					pbs_pmix_register_client(pjob, tvnodeid, &envp);
+#endif
+					ptask = momtask_create(pjob);
+					if (ptask != NULL) {
+						strcpy(ptask->ti_qs.ti_parentjobid, jobid);
+						ptask->ti_qs.ti_parentnode = myvnodeid;
+						ptask->ti_qs.ti_myvnode = tvnodeid;
+						ptask->ti_qs.ti_parenttask = fromtask;
+						if (task_save(ptask) != -1) {
+							ret = start_process(ptask, argv, envp, false);
+							if (ret == PBSE_SYSTEM) {
+								ptask->ti_qs.ti_status = TI_STATE_EXITED;
+							}
+						}
+					}
+					/* Save off the taskid to send all at once */
+					temptaskid[index] = ptask->ti_qs.ti_task;
+					tempnid[index] = tvnodeid;
+					index++;
+				} else {
+					/*
+					 ** Sending to another MOM.
+					 */
+					/* 
+					** Have we already added a stream for this mom?
+					** We don't want to open another stream to the same mom.
+					*/	
+					for (k = 0; k < momcnt; k++) {
+						if (tvnodeid == streamlist[k]) {
+							alreadyadded = 1;
+							break;
+						}
+					}
+
+					/* Open a stream to the mom if a stream isn't already open */
+					/* The host of the current vnode is phost, use phost */
+					if (phost->hn_stream == -1)
+						phost->hn_stream = tpp_open(phost->hn_host, phost->hn_port);
+					phost->hn_sister = SISTER_EOF;
+					if (phost->hn_stream < 0) {
+						sprintf(log_buffer, "tpp_open failed on %s:%d",
+							phost->hn_host, phost->hn_port);
+						log_joberr(-1, __func__, log_buffer, jobid);
+						continue;
+					}
+				
+					if (pbs_conf.pbs_use_mcast == 1 && !alreadyadded) {
+						/* add each of the tpp streams to the tpp mcast channel */
+						if (tpp_mcast_add_strm(mtfd, phost->hn_stream, FALSE) == -1) {
```

Comment:
U can call tpp_mcast_add_strm( mtfd, stream, unique) - the third parameter can be set to true which allows tpp itself to check if the stream was already added or not. So the fd will be added if the stream was not already added (if unique was set to true)